RecoMuon/MuonIdentification/plugins/MuonReducedTrackExtraProducer.h

```diff
@@ -0,0 +1,49 @@
+#ifndef RecoMuon_MuonIdentification_MuonReducedTrackExtraProducer_H
+#define RecoMuon_MuonIdentification_MuonReducedTrackExtraProducer_H
+
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "DataFormats/MuonReco/interface/MuonFwd.h"
+#include "DataFormats/TrackReco/interface/TrackExtraFwd.h"
+#include "DataFormats/Common/interface/DetSetVectorNew.h"
+#include "CommonTools/Utils/interface/StringCutObjectSelector.h"
+#include "DataFormats/TrackingRecHit/interface/TrackingRecHitFwd.h"
+#include "DataFormats/Common/interface/Association.h"
+
+class SiPixelCluster;
+class SiStripCluster;
+
+namespace pat {
+  class Muon;
+}
+
+class MuonReducedTrackExtraProducer : public edm::stream::EDProducer<> {
```

Comment:
Couldn't this become a `global` module?

Reply:
Are you sure StringCutObjectSelector is thread-safe?

Reply:
Uhm, there are examples of global producers in which there is a  `StringCutObjectSelector` as a class member. But honestly I'm not sure it is thread safe, and browsing around I also found authoritative statements that claim it is not.
In any case, I was just guessing: no reason to spend time in investigating it if it doesn't turn out immediate

Reply:
Thinking at it a bit more, I don't see why StringCutObjectSelector as used here cannot be thread safe.
@Dr15Jones , what do you think?

Reply:
I'm not confident StringCutObjectSelector is thread-safe, especially since it calls down into ROOT. We know it to be thread-friendly.

Reply:
Ok, thank you @Dr15Jones 
By the way, do you see a possible issue in (e.g.) ​PhysicsTools/​PatAlgos/​plugins/​PATLostTracks.cc?

Reply:
In `PATLostTracks` the `StringCutObjectSelector` is make in the constructor and is set to be _non lazy_ which avoids calling ROOT's cling during event processing. Unfortunately one part of the Selector system isn't thread safe: ExpressionVar. It has two `mutable` method it modifies during expression evaluation:

https://github.com/cms-sw/cmssw/blob/master/CommonTools/Utils/src/ExpressionVar.h#L26-L27